src/ForwardIterator.ts

```diff
@@ -0,0 +1,191 @@
+import { compare, Time, add as addTime } from "@foxglove/rostime";
+import Heap from "heap";
+
+import { IBagReader } from "./IBagReader";
+import { ChunkInfo, MessageData } from "./record";
+import { IteratorConstructorArgs, Decompress, ChunkReadResult } from "./types";
+
+class ForwardIterator {
+  private connectionIds?: Set<number>;
+  private heap: Heap<{ time: Time; offset: number; chunkReadResult: ChunkReadResult }>;
+  private position: Time;
+  private decompress: Decompress;
+  private chunkInfos: ChunkInfo[];
+  private reader: IBagReader;
+
+  private cachedChunkReadResults = new Map<number, ChunkReadResult>();
+
+  constructor(args: IteratorConstructorArgs) {
+    this.position = args.position;
+    this.decompress = args.decompress;
+    this.reader = args.reader;
+    this.chunkInfos = args.chunkInfos;
+
+    // if we want to filter by topic, make a list of connection ids to allow
+    if (args.topics) {
+      const topics = args.topics;
+      this.connectionIds = new Set();
+      for (const [id, connection] of args.connections) {
+        if (topics.includes(connection.topic)) {
+          this.connectionIds.add(id);
+        }
+      }
+    }
+
+    this.heap = new Heap((a, b) => {
+      return compare(a.time, b.time);
+    });
+  }
+
+  // Load the next set of messages into the heap.
+  private async loadNext(): Promise<void> {
+    let stamp = this.position;
+
+    // These are all chunks that we can consider for iteration.
+    // Only consider chunks with an endTime after or equal to our position.
+    // Chunks before our position are not part of forward iteration.
+    let candidateChunkInfos = this.chunkInfos.filter((info) => {
+      return compare(info.endTime, stamp) >= 0;
+    });
```

Comment:
It could be an improvement to sort the chunks once when creating the iterator and then keep trimming down the candidate list as chunks are consumed so there are fewer to consider as the iterator is consumed. I'll see if anything good jumps at me when making some other changes from the PR - if the changes prove to make the code harder to reason about I might leave this scan in place for now.